src/tactic/compute_degree.lean

```diff
@@ -172,15 +299,244 @@ by compute_degree_le
+`get_lead_coeff/resolve_coeff` to scan the expression of `f` and producing a hopefully simpler
+expression.  After this, it calls on `simp only [a few lemmas]` and `norm_num` to simplify further.
+If the optional `with h` argument is given, then `h` is the name that `simp_coeff` assigns to
+possible side-goals.
+-/
+meta def simp_coeff (na : parse with_ident_list) : tactic unit :=
+do t ← target >>= instantiate_mvars,
+  (t_is_eq, f, m) ← match t with
+    | `(coeff %%f %%m = _) := return (tt, f, m)
+    | `(coeff %%f %%m ≠ _) := return (ff, f, m)
+    | _ := fail "Goal is not of the form `f.coeff n = x` or `f.coeff n ≠ x`"
+    end,
+  d_nat ← guess_degree' f,
+  m_nat ← eval_expr' ℕ m,
+  guard (d_nat = m_nat) <|> fail!(
+  "`simp_coeff` checks that the expected degree is equal to the degree appearing in `coeff`\n" ++
+  "the expected degree is `{d_nat}`, but you are asking about the coefficient of degree `{m_nat}`"),
+  c_c ← poly_and_deg_to_equation f m na,
+  interactive.swap,
+  if t_is_eq then refine ``(eq.trans %%c_c _) else refine ``(ne_of_eq_of_ne %%c_c _),
+  try $ tactic.clear c_c,
+  refine ``(one_ne_zero) <|> interactive.swap,
+  resolve_coeff
+
+/--  `compute_degree` tries to close goals of the form `f.(nat_)degree = d`.  It converts the
+goal to showing that
+* the degree is at most `d`, calling `compute_degree_le` to solve this case;
+* the coefficient of degree `d` is non-zero, calling `simp_coeff` to simplify this goal.
+
+Unless the polynomial is particularly complicated, `compute_degree` either succeeds of leaves
+a simpler goal to prove.  Continue reading for a discussion of what are the current
+limitations to the tactic.
+
+`compute_degree` scans the polynomial `f` and computes the largest degree `n` that it sees.
+It then assumes that this degree `n` equals the expected degree `d`, failing otherwise.
+
+In practice, this means that if the goal is `nat_degree (X - X + 1 : ℤ[X]) = 0`, then
+`compute_degree` gives an error, since there are terms of degree `1` in the expression
+(the cancelling pair `X, -X`) and the tactic expects `1` to be the degree.  Note that this is not
+an issue if the cancellation arises in terms of degrees smaller than the maximum:
+```lean
+example : nat_degree (X ^ 2 + X - X : ℤ[X]) = 2 := by compute_degree
+```
+works. -/
+meta def compute_degree (single : parse (tk "!" )?) : tactic unit :=
+focus $ do
+  t ← target >>= (λ f, whnf f reducible),
+  match t with
+  -- the `degree` match implicitly assumes that the `nat_degree` is strictly positive
+  | `(    degree %%_ = %%n) :=
+    refine ``((degree_eq_iff_nat_degree_eq_of_pos (by norm_num : 0 < _)).mpr _)
+  | `(nat_degree %%_ = %%_) := do
+    wks ← try_core single_term_suggestions,
+    match wks with
+    | some wor :=
+      if single.is_some then refine wor.2
+      else (do pwks ← pp wor.1, fail!"Try this: exact {pwks}\n\nor\n\nTry this: compute_degree!")
+    | none := skip
```

Comment:
I agree with your interpretation of the code!

I am not entirely sure about the `focus1` split. I think that the first contradiction `(%%n0 rfl).elim` handles the case in which the exponent `n` really was `0` and you now have a possibly false goal and `%%n0` is the hypothesis `0 ≠ 0`. If you comment out this branch, then this test
```lean
example {p0 : p.nat_degree = 0} :
 (p ^ n).nat_degree ≤ 0 :=
by cases n; compute_degree_le
```
fails.

The final branch (`nat.div_self`) might be a residue from a previous implementation of `guess_degree`, where I gave special support to the degree of `X ^ n` and assigned it degree `n` instead of `n * 1`. I tried commenting it out and no current test fails. I will see if I can find a test that uses that branch, but otherwise I am happy to prune the code!